Echelle

emcangi opened the pull request:
## Pull Request Checklist
 - [X] Code Follows PEP8 guidelines (explain any exceptions)
Some lines are over length to avoid breaking in awkward positions

 - [x] New routines don't duplicate existing functionality
 - [x] Original authors of modified routines have been added as reviewers

planetarymike commented on maven_iuvs/fits_processing.py:
```diff
@@ -0,0 +1,498 @@
+
+    Parameters
+    ----------
+    file_index : index file (.npy) 
```
I think the loaded index file is an array, not an .npy filename, so this might be confusing... I guess we could define a new datatype, but let's not if we can get away with it.

emcangi commented:
We could, I just figured it made sense to agnosticize it now. Would you prefer it that way?

planetarymike commented:
Right now I think that the index generation routines are agnostic to the contents of the directories, but the echelle routines that look at the index to generate the reports and file pairing don't have any protections to make sure that the files in the index are only echelle files. Until we add protections in one or both places I think the safer option is to keep all of the index generation stuff in the echelle submodule.

emcangi commented:
For now I just moved those functions back to the echelle file. What do you think?

planetarymike commented:
I think to be consistent all of the functions that are specific to echelle and the file index need to be in echelle.py . Same for functions like get_countrate_diagnostics, which is really only useful for the echelle with the pixel ranges specified.

I think the geometry functions and the average pixel countrate functions might belong in geometry.py and statistics.py / integration.py ? 

Some of the binning stuff could go into a new binning.py file? Or does another organization scheme make more sense to you?

emcangi commented:
Yeah, I've been having difficulty deciding how to organize since a lot of the other functions are older, the module names very general ("instrument.py"), or the functions are specific to certain data levels (e.g. l1b). 

As for the index, at this point the update_index function doesn't care what's in the folder, which doesn't seem to me to be directly tied to echelle, it's more of a folder/fits parsing function. If you insist it be part of echelle, I will put it there though. get_countrate_diagnostics should be in there though, that one I forgot about.

Overall I thought it would be reasonable to upload these new functions now and then make a more concerted effort to organize next. My instinct would be to have l1a functions and l1b functions separate, but that will probably lead to duplicated code, so the right answer is probably to sort by topics ("binning", "line fitting", "geometry" etc) and then divide those files into sections if there need to be separate functions for l1b/l1a. But I can also organize them first instead.

emcangi commented:
Hmmm, yeah, let's look at it next week.

planetarymike commented:
Any reason you prefer alphabetical order? A different order would probably be more readable, but alphabetical is probably better than what we have now.

https://stackoverflow.com/questions/10289461/what-is-a-good-way-to-order-methods-in-a-python-class

emcangi commented:
Mostly because I don't know what other order would be logical, and through use of the routines, we become familiar with them and remember names of ones we use, so we can find them quickly.

Open to other ordering ideas though! Could also maybe include an index comment string at the top, but I don't know offhand if that's discourged in PEP8 or 257 or whatever they call the style now.

planetarymike commented:
The link I shared has some decent suggestions; put the most important stuff first, followed by helper methods; group related routines by subject matter. We don’t need to be strict about this but alphabetical seems cumbersome to maintain…

emcangi commented:
Oh thanks, sorry I missed it due to distractions. will have a look later.

Sent from Proton Mail mobile

-------- Original Message --------
On Aug 8, 2023, 6:38 PM, Mike Chaffin wrote:

> The link I shared has some decent suggestions; put the most important stuff first, followed by helper methods; group related routines by subject matter. We don’t need to be strict about this but alphabetical seems cumbersome to maintain…
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/lasp/maven_iuvs/pull/54#issuecomment-1670489491), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ABB36LVTXGJB23DGV3QALDLXULLZJANCNFSM6AAAAAAY7PWZSI).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

emcangi commented:
Fixes https://github.com/lasp/maven_iuvs/issues/43

planetarymike commented:
Hey Eryn,

You marked this as ready for review yesterday, but pushed some additional changes afterward. Are you done modifying these routines / should I proceed to review?

Mike

planetarymike commented:
Hi Eryn,

Given the room setup here at VEXAG / Venus as a System, I don’t think I’ll be able to review these changes this week: hopefully I can review and merge them next week at the EMUS meeting. Sorry I didn’t get to them last Friday!

Mike

emcangi commented:
I think the large ones you're looking at are commits where I organized the functions in the file according to the suggestion of top-level at the top and helper functions/lower level at the bottom. 

Not much has changed in download.py or geometry.py. In download.py, I added a few lines to correct a problem where rsync would silently fail if the local path had a space in it. In geometry I added four routines overall. In looking back at the commits, I wasn't being specific enough, thinking the changes would be evident from the commit, but I can see where this is not the case so future commit messages will be more precise. I could possibly edit the existing commit messages, although it may confuse the issue more.

planetarymike commented:
I think we need to split out changes that are strictly reorganization from content changes in a way that makes the diffs legible... I believe you that there are no major changes to the code files whose diffs are full of supposed changes, but it's impossible to verify this from the diffs. Any thoughts?

emcangi commented:
Yes, that seems like a good idea. To me the diffs are pretty clear using Sublime Merge but I bet they are annoying in the terminal. Let me look into how to remove the relevant commits...

planetarymike commented:
I did some additional reading on this topic and I think the problem is on my end— git diff can’t do structured diffs that would detect this kind of function moves, so reviewing changes like this in github will never work. I’ll spin up some more comprehensive git tools on my end and check if that resolves my ability to review this effectively.

planetarymike commented:
OK, I am now set up with a better git toolchain. The diffs are more legible, but it's still impossible to tell whether code was changed when the functions were moved--- is this the case on your end too? Clearly indicating when code was moved in a file turns out to be pretty much impossible without function level diffing, which might not exist (I couldn't find a utility to do this).

So, in the future we need to be very careful about this kind of reorganization to make merging pull requests manageable. For this PR, I think we should put unchanged code back in its original order in the files. We could consider reorganizing the code in another PR, but if it's done within the same file we'll just end up with the same problem. Thoughts?

emcangi commented:
I've investigated this today; the reorganization code should definitely be separate from the content changes. This was my mistake for not realizing this was a best practice and thinking I needed to implement changes as we discussed them up-thread. Additionally, I see now partially why the diffs are hard to read, even in Sublime Merge---some text that is common between functions, such as the "Parameters" and "Returns" headers in docstrings, don't get marked as additions or deletions or moves because they are the same between functions, even when I've moved functions wholesale. 

To my eye there are four seriously problematic commits (all the rest in the PR are either not polluted with both reorgs and content changes, or the pollute is really minimal):

- d2211f01 - removed some geometry-related functions from `fits_processing.py`; adds the function `get_pix_range`. 
- 70c525bc - Added the geometry-related functions removed from `fits_processing.py` to `geometry.py` and reorganized the functions, while also adding the new functions `make_tangent_lat_lon_plot`, `make_SCalt_plot`, `make_sza_plot`
- 8349f046 - Reorganized echelle.py and added some code to draw slit guidelines on the quicklooks in `echelle.py` and `line_fit_plot.py`. 
- f797b8d1 - Reorganized several files but no substantial content change: 
	- `download.py`
	- `geometry.py`
	- `integration.py`
	- `search.py`
	- `spice.py`

I think it would be difficult to do a rebase since in all of these commits, there is stuff we would like to keep and stuff we want to get rid of. (Incidentally, I don't know why d2211f01 and 70c525bc weren't the same commit since they involve movement of the same functions; I already tried to squash them in Sublime Merge, but it caused the branch to be diverged and I wasn't confident about resolving the issue so I was able to undo the squash.) 

You mentioned undoing the moves of any code that was relatively unchanged; this includes `integration.py`, `search.py`, and `spice.py`. I think the safest thing I could do to undo the reorganizations would be to `revert` these specific files within the commit f797b8d1. I believe then that the diff between f797b8d1 and the revert commit will show no changes for the files mentioned.  I'm not sure what you were using to look at the diffs, but [the --color-moved feature](https://dev.to/cloudx/how-to-color-the-moved-code-in-git-10ei) may be helpful once I revert the changes to `integration.py`, `search.py`, and `spice.py`.

What do you think of this plan? 

(Also, note I entered commit c4d827b today just in case I needed to do anything destructive on my fork of the repo, but I decided against it. It contains good and useful updates to the quicklook code though, and adding it shouldn't cause any additional complication to this issue)

emcangi commented:
Ok, I've now done my best to revert the problematic commits which were polluting the diffs. Have a look and let me know if we can merge this!

emcangi commented:
Ok, I've now done my best to revert the problematic commits which were polluting the diffs. Have a look and let me know if we can merge this!

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,1292 @@
+    # collect all the indices of frames with some problem and remove an equivalent number of frames from the final matrix.
+    # Make sure it's a set (unique values only)
+    all_bad_inds = sorted(list(set([*nan_light_inds, *bad_light_inds, *bad_dark_inds])))
+    dark_subtracted = np.delete(dark_subtracted, all_bad_inds, axis=0)
```
better to return NaN for invalid frames, that way return value always has same dimensions at input light array (rule of least surprise)

emcangi replied:
Fixed by ae415e4a60f788a69c1d123945a1e143fd23e89a

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,1292 @@
+    Parameters
+    ----------
+    selected_l1a : list of dictionaries
+                   selected l1a metadata entries to use
```
(presumably a list of light files)

emcangi replied:
Clarified this docstring

emcangi replied:
Fixed by d81b483549da109ff45b25607960cdaf97a7345b

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,1292 @@
+    True or False
+    """
+
+    return 'dark' in fidx['name']
```
need to add a TODO to impose a stricter test of "actually a dark"

emcangi replied:
Added

emcangi replied:
Fixed by d81b483549da109ff45b25607960cdaf97a7345b

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,1292 @@
+# Count rates ----------------------------------------------------------
+
+
+def get_avg_pixel_count_rate(hdul, spapixrange, spepixrange, return_npix=True):
```
probably this should be a generic routine in a different file?

emcangi replied:
I didn't write this one or put it in here, but happy to move it. How about moving it to geometry?

planetarymike replied:
it probably belongs in 'integration' if it belongs anywhere; 'geometry' is specifically routines for handling spacecraft and pointing location information.

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,1292 @@
+
+    Parameters
+    ----------
+    hdul : astropy FITS HDUList object
```
missing/incomplete docstring

emcangi replied:
Fixed by d81b483549da109ff45b25607960cdaf97a7345b

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,1292 @@
+
+    Returns
+    -------
+    None -- just updates the index files 
```
incorrect return value

emcangi replied:
Fixed by d81b483549da109ff45b25607960cdaf97a7345b

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,1292 @@
+    ----------
+    the_dir : string
+              path to directory containing observation data
+    new_files_limit : 
```
number of new files to add before stopping and saving the resulting index.

emcangi replied:
Fixed by d81b483549da109ff45b25607960cdaf97a7345b

planetarymike commented on maven_iuvs/file_classes.py:
```diff
@@ -29,7 +29,7 @@ def __init__(self, filename):
         self.__check_input_is_iuvs_data_filename()
         hdulist = HDUList.fromfile(filename, mode='readonly',
-                                   memmap=True, save_backup=False,
+                                   memmap=False, save_backup=False, # Note: Changed memmap to False because having it True precluded making quicklooks. --EMC 8/10/23
```
We could stop using this file_class object... it seriously slows down find_files and isn't super helpful for my workflow... immaterial to the present PR but interested in your thoughts.

emcangi replied:
I've used it in a few places, but I have no huge attachment. It is nice that it has some methods to obtain certain ephemera/data easily but it's also not enormously important. I'd like to table it for now or at least introduce a new commit to remove it, as long as that wouldn't bother anyone...

planetarymike replied:
Happy to change all the loading routines to use fits.open() instead; IUVSFITS adds a lot of overhead processing time.

emcangi replied:
Done in ced635c7bcaa29b50074b1e9d4643fb3929406f1, there weren't that many in use, mostly I had used it in echelle.py

planetarymike commented on maven_iuvs/fits_processing.py:
```diff
@@ -0,0 +1,273 @@
+    return pixrange
+
+    
+def update_index(rootpath, new_files_limit_per_run=1000):
```
needs to be moved to echelle unless this is going to be made generic for all FUV/MUV files

emcangi replied:
Fixed by 6467166340944ca6a69510b1a6725e519a1a501e

planetarymike commented on maven_iuvs/fits_processing.py:
```diff
@@ -0,0 +1,273 @@
+from maven_iuvs.search import find_files, get_latest_files
+
+
+def get_binning_scheme(hdul):
```
most of the routines in this file probably belong in `binning.py', 'fits_processing' seems too generic... misfits can go into misc

emcangi replied:
Fixed by 6467166340944ca6a69510b1a6725e519a1a501e

planetarymike commented on maven_iuvs/geometry.py:
```diff
@@ -941,3 +941,178 @@ def pixelcorner_avg(pixel_x, pixel_y, pixel_z=None,
+    return (nanfrac < 1.0)
+
+
+def find_files_missing_geometry(file_index, show_total=False):
```
echelle indexing files need to be specific to echelle

emcangi replied:
Fixed by c361af4d4d7026605a880d550706d37d7c4378c2

planetarymike commented on maven_iuvs/geometry.py:
```diff
@@ -941,3 +941,178 @@ def pixelcorner_avg(pixel_x, pixel_y, pixel_z=None,
+    return with_geom
+
+
+def make_sza_plot(ax, fitfile, linecolor="cornflowerblue"):
```
move to graphics

emcangi replied:
Fixed by c361af4d4d7026605a880d550706d37d7c4378c2

planetarymike commented on maven_iuvs/graphics/line_fit_plot.py:
```diff
@@ -3,36 +3,61 @@
     myfits : IUVSFITS or HDUList
-        IUVS FITS file to be plotted.
+             IUVS FITS file to be plotted.
+    image_to_plot : optional data override to use when plotting.
```
this probably indicates a need to split routines up: one for plotting an input array, one to extract the data from a fits file and pass is to the plotting routine. Recommend refactoring.

emcangi replied:
Fixed by 19ff4ea22866d14108ef61c8ecf5b25a8d7f81d7

planetarymike commented on maven_iuvs/graphics/line_fit_plot.py:
```diff
@@ -3,36 +3,61 @@
     integration : int
-       Integration number to be plotted.
+                  Integration number to be plotted.
+    draw_slit_lines: boolean   
```
needs to have options for FUV/MUV and echelle

emcangi replied:
Can we table this for a future commit/PR? I haven't interacted with those datasets for a while so I don't have them downloaded right now for experimenting.

planetarymike commented on maven_iuvs/graphics/line_fit_plot.py:
```diff
@@ -3,36 +3,61 @@
+           Colormap number to use from idl_colorbars package.
+    titletext : string
+                optional title to add to the image.
+    fontsizes: string 
```
give example input?

emcangi replied:
Fixed by 713fe2542d534540d67550f08173bd1e21f6abfc

planetarymike commented on maven_iuvs/graphics/line_fit_plot.py:
```diff
@@ -50,72 +75,100 @@ def detector_image(myfits, integration=0,
     cmap = idl_colorbars.getcmap(cmap)
 
-    ax.set_xlim([0, 1024])
-    ax.set_ylim([0, 1024])
```
the original intent here was to plot the full extent of the detector and overplot the file data range... this functionality needs to be preserved.

emcangi replied:
Fixed by 19ff4ea22866d14108ef61c8ecf5b25a8d7f81d7

planetarymike commented on maven_iuvs/graphics/line_fit_plot.py:
```diff
@@ -3,36 +3,61 @@
+                   fig=None, ax=None, norm=None, cmap=109, titletext="", fontsizes="medium",
+                   scale="linear", print_scale_type=True, show_colorbar=True, show_cbar_lbl=True, labels_off=False, 
+                   arange=None, prange=None, force_vmin=None, force_vmax=None):
+    """Makes a density plot of detector DN for the input FITS file and
```
General concern that the scale of the changes here will break an unspecified amount of my existing code in various notebooks... in future it would be better to add new routines that preserve existing functionality, factoring out code that needs reuse, and introducing new routines for new use cases.

emcangi replied:
Fixed by 19ff4ea22866d14108ef61c8ecf5b25a8d7f81d7

planetarymike commented on maven_iuvs/graphics/line_fit_plot.py:
```diff
@@ -50,72 +75,100 @@ def detector_image(myfits, integration=0,
-        if arange is None:
-            arange = [np.percentile(data, prange[0]),
-                      np.percentile(data, prange[1])]
+        # Range decisions used to be here, but now it is done outside in the primary plotting routine
```
this is definitely a breaking change and should be handled differently...

emcangi replied:
Fixed by 19ff4ea22866d14108ef61c8ecf5b25a8d7f81d7

planetarymike commented on maven_iuvs/graphics/line_fit_plot.py:
```diff
@@ -50,72 +75,100 @@ def detector_image(myfits, integration=0,
         if scale == "linear":
-            norm = mpl.colors.Normalize(vmin=arange[0],
-                                        vmax=arange[1])
+            norm = mpl.colors.Normalize(vmin=vmin, vmax=vmax)
```
In future, please avoid stylistic changes that don't affect the content of the code or fix PEP8 violations--- this kind of thing makes it much harder to read diffs. The rule is to keep the style of the original code unless there are substantive changes. This happens in lots of places, not just here.

planetarymike commented on maven_iuvs/integration.py:
```diff
@@ -255,8 +255,17 @@ def fit_line(myfits, wavelength,
-            DN = myfits['detector_dark_subtracted'].data[iint, ispa]
-            DN_unc = myfits['Random_dn_unc'].data[iint, ispa]
+
+            try:
```
better to read the file identifier and select the correct treatment for l1a or l1b files...

emcangi replied:
Fixed by 1669a94a7aa155787effe1d66b032aecf75395d3

planetarymike commented on maven_iuvs/integration.py:
```diff
@@ -255,8 +255,17 @@ def fit_line(myfits, wavelength,
+            try: 
+                DN_unc = myfits['Random_dn_unc'].data[iint, ispa] # l1b
+            except:
+                DN_unc = np.zeros_like(DN) # l1a -- this is wrong obviously
```
better to fill with NaN

emcangi replied:
Fixed by 1669a94a7aa155787effe1d66b032aecf75395d3

planetarymike commented on maven_iuvs/miscellaneous.py:
```diff
@@ -51,3 +51,34 @@ def mirror_dn_to_deg(dn, inverse=False):
+    return idx, array[idx]
+
+
+def get_grad_colors(L, cmap, strt=0, stp=1, mikes=False):
```
should be in graphics

emcangi replied:
Fixed by 6467166340944ca6a69510b1a6725e519a1a501e

planetarymike reviewed (commented):
Hi Eryn,

Thanks for making this much more legible--- I left some comments, let me know what you think.

Mike

emcangi commented on maven_iuvs/echelle.py:
```diff
@@ -8,6 +8,7 @@
 import matplotlib.transforms as transforms
 from astropy.io import fits
 import math 
+from tqdm.auto import tqdm
```
tqdm has been added to setup.py, thanks for the catch!

emcangi replied:
Fixed by ecdef50a6424c0982fa48d2380b0896f03bd4539

emcangi commented on maven_iuvs/constants.py:
```diff
@@ -1,3 +1,5 @@
 
 R_Mars_km = 3.3895e3  # [km]
+
+nan_color = "#424242"
```
Actually, I removed it entirely because there is already a variable in `graphics.py` called `color_dict` (which I missed) that contains a gray that is close enough. Replaced instances of `nan_color` with `color_dict['darkgrey']`.

emcangi replied:
Fixed by 2e510edc07dda7e866884b8f11f1cf76892d0571

emcangi commented on maven_iuvs/download.py:
```diff
@@ -131,11 +130,13 @@ def setup_user_paths():
+ 
 def get_default_data_directory(level='l1b'):
-    """Returns default l1b_dir defined in user_paths.py, creating the
+    """
```
Yes, the original code (perhaps written by you or Kyle or both) contained the summary lines starting on the same line as the opening triple quote (mostly, I think). In some of my functions I started the summary on the next line because that looks better to me without really thinking about it/checking what PEP actually says. In undoing some of the messy commits, I tried to move these back to their original location on the same line as """ to ensure line numbers would match so that the overall diff would be cleaner for you [and others] to read. 

I did just check PEP and it says that either is acceptable:

> The summary line may be on the same line as the opening quotes or on the next line.

I'm currently using SublimeLinter with flake8 conventions. We can discuss to agree on a convention and then I can make a new commit later that just focuses on cleaning up any formatting issues in the docstrings/other comments!

planetarymike replied:
Let's open an issue on this

emcangi replied:
Issue https://github.com/lasp/maven_iuvs/issues/58 is opened, resolving this conversation

emcangi commented on maven_iuvs/download.py:
```diff
@@ -482,11 +483,6 @@ def sync_data(spice=True, level='l1b',
         else:
             raise ValueError('local_dir must be l1a or l1b')
 
-    # Check to be sure that the vm is in the known_hosts file.
```
I've looked into it and I have no idea. I think this must be a mistake--I've added it back in since it should definitely be there.

emcangi replied:
Fixed by bf5ed3a31fdd88b6eb6eb728dc473e3f0bda45ef

emcangi commented on maven_iuvs/download.py:
```diff
@@ -558,6 +555,9 @@ def sync_data(spice=True, level='l1b',
             # save the files to rsync to temporary files
             # this way rsync can use the files_from flag
             transfer_from_production_file = tempfile.NamedTemporaryFile()
+            # save files to sync to a local file that can be accessed for troubleshooting
```
Oh, these comment lines were used when I was still diagnosing #56. I've removed these lines.

emcangi replied:
Fixed by bf5ed3a31fdd88b6eb6eb728dc473e3f0bda45ef

emcangi commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,1292 @@
+            print('  No issues.')
+
+
+# QUICKLOOK CODE ======================================================
```
Created graphics/echelle_graphics.py, can rename to echelle.py if desired but thought that it might be confused with the other echelle.

emcangi replied:
Fixed by ae415e4a60f788a69c1d123945a1e143fd23e89a

emcangi commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,1292 @@
+              
+    Returns
+    ----------
+    Completed quicklook figure.
```
Updated the docstring and cleaned up the return values so that they are all strings rather than some strings and some booleans.

emcangi replied:
Fixed by d81b483549da109ff45b25607960cdaf97a7345b and ae415e4a60f788a69c1d123945a1e143fd23e89a

emcangi commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,1292 @@
+
+    Returns
+    ----------
+    selected_lights
```
Fixed by d81b483549da109ff45b25607960cdaf97a7345b

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,886 @@
+    """
+
+    # Retrieve dark frames for subtraction
+    first_dark, second_dark = get_dark_frames(dark_fits)
```
I think assuming that there is always a first and second dark will cause problems when processing the very earliest mission data, but we can deal with that later.

emcangi replied:
Right now it throws an error if there's only 1, returns the 2 if there are 2, and returns 2 with the second being an average of 2:end if there's > 2. I can change the unpacking to a tuple at a later point though

emcangi replied:
Started on this with bcddfdf576c7739e451c3c8ae3a396c4dd0193ae

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,886 @@
+    dark_subtracted, nan_light_inds, bad_light_inds, nan_dark_inds = subtract_darks(light_fits, first_dark, second_dark)
+    
+    # check that we still have valid data
+    if dark_subtracted.shape[0] == 0: # Some files may have no acceptable frames
```
exceptions on returned data should be raised inside the function that returns the data

emcangi replied:
Fixed by 1fd9de2318dba503f971986065f883b8a930c043

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,886 @@
+        raise Exception(f"Missing critical observation data: no valid darks")
+
+    # dark subtraction
+    dark_subtracted, nan_light_inds, bad_light_inds, nan_dark_inds = subtract_darks(light_fits, first_dark, second_dark)
```
my preference would be to call get_dark_frames() inside subtract_darks() unless this function (coadd_lights) needs  to know about the dark values themselves

emcangi replied:
Fixed by 8e993bb2abcc2337088f44ee0ad8e4006fedcc1f

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -0,0 +1,886 @@
+
+
+    """
+    Hlya_spapixrange = np.array([ech_slit_start, ech_slit_end])
```
need to add an exception to abort if file is not ech, since the pix ranges make no sense for FUV or MUV data

emcangi replied:
is it not enough to already be in the echelle module? Wouldn't we thus want to apply similar error handling in other functions?

planetarymike replied:
Yeah, probably fine to just trust the user in this case.

planetarymike commented on maven_iuvs/graphics/graphics.py:
```diff
@@ -1304,3 +1327,102 @@ def reset_symlog_labels(fig, axes):
             tick.set_pad(11)
+
+
+def make_sza_plot(ax, fitfile, linecolor="cornflowerblue"):
```
would be nice if ax was an optional argument so these could be called directly on a fits file, but that's not needed now.

emcangi replied:
Easy enough, fixed by 48e969aa8240f4d5c25b2e7a76aeeb796a9e5b50

planetarymike commented on maven_iuvs/graphics/line_fit_plot.py:
```diff
@@ -50,72 +146,56 @@ def detector_image(myfits, integration=0,
+        # Allows overriding the vmin and vmax; for instance, if you 
+        # have multiple externally-created plots that you want to
+        # display with a shared colorbar.
+        vmin = force_vmin if force_vmin is not None else arange[0] 
```
seems like arange and force_vmin,max are synonymous?

emcangi replied:
Yes, must be an artifact, fixed by 07835d00fd301d96b8d508a211c52cee7710a049

planetarymike commented on maven_iuvs/instrument.py:
```diff
@@ -47,6 +47,9 @@
 slit. This is out of 1024 pixels (index 1023) for a 1024x1024 pixel
 detector."""
 
+ech_slit_start = 346  # Starting pixel of slit in echelle mode
```
slit position for H/D Lyman alpha

emcangi replied:
Oh yeah, thanks. You would think I would know this having looked at the images 293752 times.

emcangi replied:
Fixed by 0170b4e961f6c74d1ac2a1ea7d407fde769d4baf

planetarymike reviewed (commented):
Looks great--- I added some small comments, let me know what you think. Hoping code changes will be significantly less onerous in the future

planetarymike commented:
still working on this? happy to review now or later

emcangi commented:
I think I've addressed every comment, but it's hard to tell because of the way it displays. Take another final look?

planetarymike commented:
LGTM, thanks for persevering through these changes!

planetarymike reviewed (approved):
LGTM
